### PR TITLE
Maxlength component

### PR DESCRIPTION
**Pull request by johnbender**
Replaces the browser's `maxlength` functionality since it doesn't work consistently or sanely across browsers with respect to newlines.

### CODE REVIEW

**Comment by johnbender** [comment]
http://maxlength-component.origin.formcore.fgview.com/views/text-inputs.html

Test there on the 250 character limited `textarea` by pasting in the code.

**Comment by zachleat** [comment]
Hmm. The paste stuff here has some issues.

1. Paste when textarea already has a value but is not selected (should insert at cursor)
1. Paste when textarea already has a value and is selected (should replace and change cursor)
1. Undo/redo should still be respected?

**Comment by johnbender** [comment]
Nightmares. I'll take a look.

**Comment by johnbender** [comment]
@zachleat @toddparker 

The last one seems like a nightmare especially since undo information is time sensitive on some platforms. Do we want to try to tackle that, is the plugin fine without it, or should we just forgo the whole exercise?

**Comment by toddparker** [comment]
After the other numeric change going sideways, I want to err on the side of less risky things in formcore so I'm open to ideas on a lighter touch here.

File: js/maxlength-input.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		this.el.value = newValue;
+	};
+
+	MaxlengthInput.prototype.onPaste = function( e ){
+		var self = this;
+
+		// force the text to look right after the paste has applied
+		setTimeout(function(){
+			self.alterValue();
+		});
```
Comment (johnbender) [inline js/maxlength-input.js]: Notably I'm letting the alterations apply before replacing them to let the browser do some of the work in trimming pasted input. This seemed to help with corner cases where we're pasting into the middle of the text. In fact in my testing it truncates the pasted input as expected first. 

In the `keyDown` handler I do check if the return will put it past the length and prevent the default behavior though. 

The timeout could be moved inside `alterValue`.

File: js/maxlength-input.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+		// if any character would put us at the
+		if(this.valueLength() >= this.maxlength){
+			event.preventDefault();
+			return;
+		}
+
+		// if the user hits return and that would put the length above
+		// the max, prevent the return character addition
+		if(event.keyCode == 13 && this.valueLength() + 2 > this.maxlength){
```
Comment (johnbender) [inline js/maxlength-input.js]: Magic `2` should be length of `"\r\n"` in replace below.

**Comment by zachleat** [comment]
Found these, related to this bug: https://bugzilla.mozilla.org/show_bug.cgi?id=702296 https://bugs.webkit.org/show_bug.cgi?id=74686 Worth tracking

**Comment by zachleat** [comment]
Added a few more notes here https://github.com/scottjehl/Device-Bugs/issues/73 and I’m finally on board with this @johnbender. There is a `console.log` hiding in here but I’ll merge and remove it

**Comment by johnbender** [comment]
Oh snap my bad, thanks for taking it out.

**Pull request merged**